### PR TITLE
ffi protocol improvements 

### PR DESCRIPTION
**Pull request by theomonnom**
- allow to capture audio data without copy
- added play_from_disk example
- separate the info and owned messages 
- to_i420 destination (to avoid unnecessary allocations) 

breaking changes: 
 - capture_frames is now async and automatically handle the clock
 - NativeAudioSource::new now requires the sample_rate and the num_channels

### CODE REVIEW

File: examples/play_from_disk/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+            let ms = FRAME_DURATION.as_millis() as u32;
+            let num_samples = (header.sample_rate / 1000 * ms) as usize;
+
+            log::info!("sample_rate: {}", header.sample_rate);
+            log::info!("num_channels: {}", header.num_channels);
+            log::info!("max samples: {}", max_samples);
+            log::info!("chunk size: {}ms - {} samples", ms, num_samples);
+
+            let mut written_samples = 0;
+            while written_samples < max_samples {
```
Comment (davidzhao) [inline examples/play_from_disk/src/main.rs:162]: this looks great! so much simpler to do now

File: livekit-webrtc/src/native/audio_source.rs [hunk 2]
```diff
@@ -73,52 +76,195 @@ impl NativeAudioSource {
...
+
+        let mut inner = self.inner.lock().await;
+        let mut interval = interval(Duration::from_millis(10));
+
+        loop {
+            let Some(data) = self.next_frame(&mut inner, frame) else {
+                break;
             };
 
+            interval.tick().await;
```
Comment (davidzhao) [inline livekit-webrtc/src/native/audio_source.rs:142]: would there be clock drift issues here? i.e.
* 100ns spent in processing (between on_captured_frame and next_frame)
* sleeps for 10ms
* repeat 10 times, audio would be behind by 1ms

**Review by davidzhao — COMMENTED** [review]
Awesome work!! just one comment re: sync.

File: livekit-webrtc/src/native/audio_source.rs (hunk 2 above)
Comment (theomonnom) [reply livekit-webrtc/src/native/audio_source.rs:142]: So I now keep the same intervals between capture_frame calls, I think the tokio timer precision should be good enough for 10ms

File: examples/play_from_disk/src/main.rs [hunk 3]
```diff
@@ -146,7 +146,7 @@ async fn main() -> Result<(), Box<dyn Error>> {
     tokio::spawn({
         let room = room.clone();
         async move {
-            const FRAME_DURATION: Duration = Duration::from_millis(1000); // Write 1s of audio at a time
+            const FRAME_DURATION: Duration = Duration::from_millis(10); // Write 1s of audio at a time
```
Comment (davidzhao) [inline examples/play_from_disk/src/main.rs]: change comment as well?

File: livekit-webrtc/src/native/audio_source.rs (hunk 2 above)
Comment (davidzhao) [reply livekit-webrtc/src/native/audio_source.rs:142]: sorry, I misunderstood the `tick` part. understood now that you are keeping the same interval obj.

**Review by davidzhao — APPROVED** [review]
lgtm!

**Pull request merged**